src/global_scheduler/global_scheduler_algorithm.c

```diff
@@ -164,66 +191,132 @@ db_client_id get_photon_id(global_scheduler_state *state,
   return photon_id;
 }
 
+double inner_product(double a[], double b[], int size) {
+  double result = 0;
+  for (int i = 0; i < size; i++) {
+    result += a[i] * b[i];
+  }
+  return result;
+}
+
+/**
+ * Main new task handling function in the global scheduler.
+ *
+ * @param state Global scheduler state.
+ * @param policy_state State specific to the scheduling policy.
+ * @param task New task to be scheduled.
+ * @return  Void.
+ */
 void handle_task_waiting(global_scheduler_state *state,
                          global_scheduler_policy_state *policy_state,
                          task *task) {
   task_spec *task_spec = task_task_spec(task);
+
   CHECKM(task_spec != NULL,
          "task wait handler encounted a task with NULL spec");
   /* Local hash table to keep track of aggregate object sizes per local
    * scheduler. */
-  object_size_entry *tmp, *s = NULL, *object_size_table = NULL;
+  object_size_entry *object_size_table = NULL;
   bool has_args_by_ref = false;
+  bool task_feasible = false;
+  int64_t task_object_size = 0; /* total size of task's data. */
 
-  object_size_table =
-      create_object_size_hashmap(state, task_spec, &has_args_by_ref);
+  object_size_table = create_object_size_hashmap(
+      state, task_spec, &has_args_by_ref, &task_object_size);
 
-  if (!object_size_table) {
-    char id_string[ID_STRING_SIZE];
-    if (has_args_by_ref) {
-      LOG_DEBUG(
-          "Using simple policy. Didn't find objects in GS cache for task = %s",
-          object_id_to_string(task_task_id(task), id_string, ID_STRING_SIZE));
-      /* TODO(future): wait for object notification and try again. */
-    } else {
-      LOG_DEBUG("Using simple policy. No arguments passed by reference.");
+  /* Go through all the nodes, calculate the score for each, pick max score. */
+  local_scheduler *ls = NULL;
+  double best_photon_score = -1;
+  db_client_id best_photon_id = NIL_ID; /* best node to send this task */
+  for (ls = (local_scheduler *) utarray_front(state->local_schedulers);
+       ls != NULL;
+       ls = (local_scheduler *) utarray_next(state->local_schedulers, ls)) {
+    /* For each local scheduler, calculate its score:
+     * 1. check hard constraints first */
+    if (!constraints_satisfied_hard(ls, task_spec)) {
+      continue;
     }
-    UNUSED(id_string);
-    handle_task_round_robin(state, policy_state, task);
-    return;
-  }
+    /* This node satisfies the hard capacity constraint. Calculate its score. */
```

Comment:
All of the logic about computing a "score" should be separated out as a different function so we can easily change the computation of the score without changing how the scheduling policy uses the score (or so we can use the score in another scheduling policy).